feat(state): Update `Proposal` with `BlobID`

alesforz opened the pull request:
Part of the work of https://github.com/berachain/cometbft/pull/26.

### Context
A `Proposal` represents a block proposal that peers vote on during consensus. To be valid, it must be signed by the correct proposer for the given height and round.

A `Proposal` includes a `BlockID`, which uniquely identifies the proposed block. This `BlockID` contains the Merkle root hash of the block’s parts, ensuring integrity as the block is split and gossiped across the network.

Since blocks can now have an associated blob, we introduce a unique identifier for the blob in the `Proposal` struct. This allows peers to correctly associate a blob with its corresponding block during gossip. Additionally, because the `Proposal` is signed, this prevents malicious actors from altering or replacing the blob linked to a specific block.

### Changes
This PR extends the `Proposal` struct and its methods with a `BlobID` field.

Commit:
Extend type `Proposal` with `BlobID` field

Commit:
Update `Proposal` methods `ValidateBasic()` and `String()` to use `BlobID`

Commit:
Add `BlobID` to `Proposal` protobuf.

Regenerated the proto files as well.
Moved the proto definition of `BlobID` to the `types.proto` file to fix an
import cycle.

Commit:
Add function to canonalize `BlobID`

Commit:
Update `Proposal` to/from protobuf functions to include `BlobID`

Commit:
Update `Proposal` unit tests to use `BlobID`

Commit:
Add method `IsNil()` to type `BlobID`

Commit:
`State.defaultDecideProposal() assigns an empty `BlobID` to `Proposal`.

Since `BlobID` is not used in the consensus logic yet, this change is harmless.
However, it allows us to keep this PR focused solely on `Proposal`-related
changes.

In a future PR, we’ll handle the proper creation of `BlobID` by implementing
blob splitting into multiple parts.

Additionally, this PR includes a refactor for improved readability.

Commit:
Add missing `BlobID` param to `decideProposal()`

Commit:
Add missing `BlobID` param in calls to `NewProposal()` in state_test.go

Commit:
Add missing `BlobID` param in calls to `NewProposal()` in replay_test.go

Commit:
Add missing `BlobID` param in calls to `NewProposal()` in pbts_test.go

Commit:
Add missing `BlobID` param in calls to `NewProposal()` in byzantine_test.go

Commit:
Fixed unit test TestConsMsgsVectors.

The `Proposal` sub-test was using an incorrect hex string to check the test's
results.
The string changed because we added the `BlobID` field to the `Proposal` type,
thus changing its encoding to protobuf.

Also rearrenged the code for better readability.

Commit:
Fixed unit test `TestWALEncoder`.

It was using an incorrect hex string to check the test's results.
The string changed because we added the `BlobID` field to the `Proposal` type,
thus changing its encoding to protobuf.

Also rearrenged the code for better readability.

Commit:
Fixed unit test `TestPrivvalVectors`.

The sub-tests "Proposal Request", "Proposal Response", and "Proposal Response
with error" were was using an incorrect hex string to check the test's results.

The string changed because we added the `BlobID` field to the `Proposal` type,
thus changing its encoding to protobuf.

Also rearrenged the code for better readability.

alesforz commented:
The linter is failing, but that's not a real problem.
It's complaining about not using the blob object, but in the next PR(s) the code will use it.

Commit:
Extend `EventDataCompleteProposal` with `BlobID` field

Commit:
`Hash()` method of type `Blob` checks return `nil` if empty blob.

jmalicevic commented on internal/consensus/state.go:
```diff
@@ -1310,11 +1335,13 @@ func (cs *State) isProposalComplete() bool {
 	if cs.Proposal == nil || cs.ProposalBlock == nil {
 		return false
 	}
+
```
```suggestion
```

jmalicevic commented on internal/consensus/state.go:
```diff
@@ -1310,11 +1335,13 @@ func (cs *State) isProposalComplete() bool {
 	if cs.Proposal.POLRound < 0 {
 		return true
 	}
+
```
```suggestion
```

jmalicevic commented on types/canonical.go:
```diff
@@ -33,6 +34,25 @@ func CanonicalizeBlockID(bid cmtproto.BlockID) *cmtproto.CanonicalBlockID {
+		panic(fmt.Sprintf("error parsing BlobID from protobuf: %v", err))
+	}
+
+	if blobID.IsNil() {
```
Looking at how this is done for blocks , I see that there is a check `blobID == nil` . Not sure what the diff is but maybe we want to include the same check here? 
```suggestion
	if blobID == nil || blobID.IsNil() {
```

alesforz replied:
`BlobIDFromProto()` returns a `BlobID` not a `*BlobID`, therefore the `blobID` variable cannot be `nil`. In contrast, `BlockIDFromProto()` returns  a `*BlockID`, therefore the code must check if it's `nil` before calling the `IsNil()` method to prevent a panic.

So in this case we don't need the `nil` check.

jmalicevic commented on types/proposal.go:
```diff
@@ -114,16 +132,20 @@ func (p *Proposal) IsTimely(recvTime time.Time, sp SynchronyParams) bool {
 		p.Round,
 		p.BlockID,
 		p.POLRound,
+		p.BlobID,
```
if `p.BlobID` is nil, will it output anything or cause an error?

alesforz replied:
`BlobID` isn't a pointer, so it won't be `nil`. In any case, if it were `nil`, Go would just print "nil".

jmalicevic commented on types/proposal.go:
```diff
@@ -149,17 +171,19 @@ func (p *Proposal) ToProto() *cmtproto.Proposal {
+		PolRound:  p.POLRound, // FIXME: names do not match
+		Timestamp: p.Timestamp,
+		Signature: p.Signature,
+		BlobID:    p.BlobID.ToProto(),
```
Should we be checking that the `BlobID` is not nil before calling `ToProto`?

alesforz replied:
`BlobID` isn't a pointer, so it won't be `nil`. The `ToProto()` method checks if the `BlobID` is "empty", and if so, it returns an empty protobuf `BlobID` (`BlockID` does the same)

jmalicevic commented on types/proposal.go:
```diff
@@ -169,20 +193,26 @@ func ProposalFromProto(pp *cmtproto.Proposal) (*Proposal, error) {
-	p.POLRound = pp.PolRound // FIXME: names do not match
-	p.Timestamp = pp.Timestamp
-	p.Signature = pp.Signature
+	blobID, err := BlobIDFromProto(&pp.BlobID)
```
Same as above, should we be checking if it is `nil` (maybe the check is performed inside `BlobIDFromProto`? ) ?

alesforz replied:
yes, `BlobIDFromProto` checks if the given argument is `nil` and returns an error if it is.

jmalicevic replied:
But can we not have a valid scenario where the blobID is nil (no blob in this proposal) and therefore it is ok that this is `nil` ? It should just be properly marshalled somehow.

alesforz replied:
The `BlobID` [field](https://github.com/alesforz/cometbft/blob/944016e304ffb62f2974119b95126fb6d042eab3/types/proposal.go#L38) in the `Proposal` struct is a concrete type, not a pointer. Therefore, it can't be `nil`. Similarly, the `BlobID` field of the 

the [`ToProto()`](https://github.com/alesforz/cometbft/blob/944016e304ffb62f2974119b95126fb6d042eab3/types/blob.go#L94) and [`FromProto()`](https://github.com/alesforz/cometbft/blob/944016e304ffb62f2974119b95126fb6d042eab3/types/blob.go#L107) methods of type `BlobID` return a concrete type, not a pointer. So I don't see how it could be set to `nil`.

jmalicevic commented:
> Update isProposalComplete() to account for the blob. We need to handle this carefully, as a missing blob could mean either that it hasn’t arrived yet or that the block simply doesn’t have one. We can distinguish between these cases by checking the Proposal: if BlobID is empty, the block has no blob. Otherwise, a blob exists, and we should wait for it.

If this is not supposed to be part of this PR, I would remove it from the description

jmalicevic commented on internal/consensus/state.go:
```diff
@@ -1284,23 +1285,47 @@ func (cs *State) defaultDecideProposal(height int64, round int32) {
 			cs.ValidRound,
 			propBlockID,
 			block.Header.Time,
+			types.BlobID{}, // TODO: add part set
```
Why is this not using the `blob` returned by `createPropsalBlock`

alesforz replied:
I want to put the splitting into parts + gossip of the blob in a separate PR not to mix things. This PR only updates the `Proposal` struct.

jmalicevic replied:
Ok, as long as this is clear then and you don't forget it :)

melekes reviewed (approved):
LGTM minus comments from Jasmina

alesforz commented:
> > Update isProposalComplete() to account for the blob. We need to handle this carefully, as a missing blob could mean either that it hasn’t arrived yet or that the block simply doesn’t have one. We can distinguish between these cases by checking the Proposal: if BlobID is empty, the block has no blob. Otherwise, a blob exists, and we should wait for it.
> 
> If this is not supposed to be part of this PR, I would remove it from the description

I was thinking whether to include this change or not in this PR 🤔 
It can fit here, but I also see it as part of the PR dealing with consensus rounds updates.

Commit:
updated docs

Commit:
Update `State.isProposalComplete()`.

Rationale: If a block has an associated blob, we set the `BlobID` in `Proposal`
after returning from `PrepareProposal()`, causing the `IsNil()` check to
return false. If the block does not have a blob, the `BlobID` field in
`Proposal` will remain its zero value, making `IsNil()` return true.

If a blob is present, `cs.ProposalBlob.IsNil()` will always return true until
`cs.ProposalBlob` is assigned a non-empty `[]byte` slice. This assignment
happens only after we have received all the blob parts and we have successfully
reconstructed the blob.

alesforz commented:
> > > Update isProposalComplete() to account for the blob. We need to handle this carefully, as a missing blob could mean either that it hasn’t arrived yet or that the block simply doesn’t have one. We can distinguish between these cases by checking the Proposal: if BlobID is empty, the block has no blob. Otherwise, a blob exists, and we should wait for it.
> > 
> > 
> > If this is not supposed to be part of this PR, I would remove it from the description
> 
> I was thinking whether to include this change or not in this PR 🤔 It can fit here, but I also see it as part of the PR dealing with consensus rounds updates.

Update, I pushed the change in this PR. See the last commit.

alesforz commented on internal/consensus/state.go:
```diff
@@ -1339,6 +1339,14 @@ func (cs *State) isProposalComplete() bool {
+		// wait for the blob before marking the proposal as complete.
+		if !cs.Proposal.BlobID.IsNil() {
+			return !cs.ProposalBlob.IsNil()
+		}
```
Rationale: If a block has an associated blob, we set the `BlobID` in `Proposal`
after returning from `PrepareProposal()`, causing the `IsNil()` check to
return false. If the block does not have a blob, the `BlobID` field in
`Proposal` will remain its zero value, making `IsNil()` return true.

If a blob is present, `cs.ProposalBlob.IsNil()` will always return true until
`cs.ProposalBlob` is assigned a non-empty `[]byte` slice. This assignment
happens only after we have received all the blob parts and we have successfully
reconstructed the blob.
